### PR TITLE
[SYCL][Fusion] API for kernel fusion library

### PR DESCRIPTION
**Pull request by sommerlukas**
This is the second patch in a series of patches to add an implementation of the[kernel fusion extension](https://github.com/intel/llvm/pull/7098). We have split the implementation into multiple patches to make them more easy to review.

This patch can be reviewed and merged independently of #7416. 

This patch adds the first components for the JIT compiler used for implementation of kernel fusion at runtime , concretely:
* API definitions
* Input translation from SPIR-V to LLVM IR
* Insertion of fused kernel function stub and metadata
* Supporting infrastructure such as compiler options.

CMake logic to link and code to invoke this JIT from the SYCL runtime will follow in a later patch.

Co-authored-by: Lukas Sommer <lukas.sommer@codeplay.com>
Co-authored-by: Victor Perez <victor.perez@codeplay.com>
Signed-off-by: Lukas Sommer <lukas.sommer@codeplay.com>

### CODE REVIEW

**Comment by sommerlukas** [comment]
The design document for the overall implementation of kernel fusion can be found [here](https://github.com/intel/llvm/pull/7204).

**Comment by Naghasan** [comment]
@bader @pvchupin Can anyone help with the review here ?

**Review by steffenlarsen — COMMENTED** [review]
Just a couple initial comments. I think I'll need a good cup of coffee for this one, so I will return in a couple of days.

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (AlexeySachkov) [inline sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp:33]:
```suggestion
    auto *ConstantMD = cast<ConstantAsMetadata>(MDOp);
```

`cast` already has embedded `assert` within it.

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (AlexeySachkov) [inline sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp:53]: Please note that there is an ongoing effort to allow less than 3 operands in `reqd_work_group_size` metadata, see #7450 (@steffenlarsen can link PRs to other components)

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (AlexeySachkov) [inline sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp:51]:
```suggestion
  if (auto *MD = KernelFunction->getMetadata(REQD_WORK_GROUP_SIZE_ATTR)) {
```

?

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (AlexeySachkov) [inline sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp:82]: If they were present in the original program, then yes, we must care about them. SYCL code may set certain values to them and during SPIR-V -> LLVM IR translation spec constants will be replaced with a regular constants. Default initialization will be used if no values provided. This replacement is a non-reversible operation, we won't preserve any information to make those spec constants re-configurable later

**Review by AlexeySachkov — COMMENTED** [review]
Just a brief review, I haven't been able to grasp it in details yet

File: sycl-fusion/common/include/Kernel.h
Comment (sommerlukas) [reply sycl-fusion/common/include/Kernel.h]: That makes sense, I reordered them, thanks.

File: sycl-fusion/README.md
Comment (sommerlukas) [reply sycl-fusion/README.md:3]: I added links to the experimental extension proposal and the design document. Note that the link for the design document will only work after #7204 has been reviewed and merged.

File: sycl-fusion/CMakeLists.txt
Comment (sommerlukas) [reply sycl-fusion/CMakeLists.txt:10]: I reported this in https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1750.

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (sommerlukas) [reply sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp:49]: Thanks for pointing this out, I wasn't aware of that. We should be able to adapt our processing of the attributes once that PR is merged, to match the new behavior.

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (sommerlukas) [reply sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp]: Yes, we definitely need to handle them. At the point where the SYCL runtime calls the JIT, the values for specialization constants should be known, so we could replace them with the user-specified value. Alternatively, we would need to make sure the spec constants are re-configurable in the output of the JIT compilation process and have the driver perform the instantiation. 

I updated the TODO to reflect this, so we can address it in a later PR.

**Comment by sommerlukas** [comment]
Thank you for the first reviews @steffenlarsen and @AlexeySachkov!

> Since this is a new library we may need to determine who would be the best reviewers going forward for these. I'm going to add a few people. @sommerlukas, can you please also update CODEOWNERS with this change (adding Codeplay folks)?

@pvchupin: I've added the Codeplay team that has been working on this implementation/library to `CODEOWNERS`. Let me know if I should add more people/teams.

File: sycl-fusion/common/include/Kernel.h
Comment (steffenlarsen) [inline sycl-fusion/common/include/Kernel.h:148]: Unless I am missing something, it seems like the implementation of this function could be moved to `getKernelFor` and `hasKernelFor` could use that directly.

File: sycl-fusion/jit-compiler/lib/KernelFusion.cpp
Comment (steffenlarsen) [inline sycl-fusion/jit-compiler/lib/KernelFusion.cpp:33]: Nit; I am not a big fan of shadowing `Err` here. Granted it is fully valid, it can be confusing if you don't notice the lambda argument shadowing it. It's a very small function though, so it's by no means blocking.

File: sycl-fusion/jit-compiler/lib/KernelFusion.cpp
Comment (steffenlarsen) [inline sycl-fusion/jit-compiler/lib/KernelFusion.cpp:116]: This function uses `auto` a lot and though the reader may have a vague idea of what some of these are, most of them wouldn't be all that more verbose with explicitly stated types and then people would know for sure what they are when they read this.

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (steffenlarsen) [reply sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp:49]: I think for this particular case https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1726 is also relevant. Note that in the reverse-translator case you will always get 3 operands back, but I believe you also care about generated LLVM IR that hasn't gone through the translator, in which case you may need to adjust for optional parameters.

**Review by steffenlarsen — COMMENTED** [review]
Just a couple more comments, but I think it looks good overall.

Note that the comment about the use of `auto` should probably also apply to a few other files. It doesn't have to be all uses of `auto`, for example `X::create` should be clear enough about the type being `X`, but for those that aren't necessarily clear about exactly what type it uses would be nice to have explicit.

Also, what are the plans for testing this? I realize much of this is just the skeleton with more to come, but is there any of this that might make sense to unit-test now or does it make more sense to wait?

File: sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp
Comment (sommerlukas) [reply sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp:49]: Thanks for the hint!

Right now, we only care about consuming SPIR-V. Untranslated LLVM IR will become relevant for targets that the DPC++ frontend currently does not compile to SPIR-V, e.g., targeting CUDA or AMD GPUs. Will keep this in mind for when we work on that.

File: sycl-fusion/common/include/Kernel.h
Comment (sommerlukas) [reply sycl-fusion/common/include/Kernel.h:143]: You're right, thanks.

File: sycl-fusion/jit-compiler/lib/KernelFusion.cpp
Comment (sommerlukas) [reply sycl-fusion/jit-compiler/lib/KernelFusion.cpp:117]: I reduced the use of `auto` in this file/function and also other implementation files.

File: sycl-fusion/jit-compiler/lib/KernelFusion.cpp
Comment (sommerlukas) [reply sycl-fusion/jit-compiler/lib/KernelFusion.cpp]: Renamed to avoid shadowing.

**Review by pvchupin — APPROVED** [review]

**Comment by sommerlukas** [comment]
Thanks for the approvals!

Can either of you please merge this PR, I'm not authorized to merge this, @steffenlarsen @pvchupin?

File: sycl-fusion/jit-compiler/include/JITContext.h
Comment (aelovikov-intel) [inline sycl-fusion/jit-compiler/include/JITContext.h:50]: We have switched to C++17 as a minimal required version mid 2022.

**Pull request merged**
